src/google/protobuf/generated_message_tctable_decl.h

```diff
@@ -391,12 +401,21 @@ struct alignas(uint64_t) TcParseTableBase {
 
   // Field entry for all fields.
   struct FieldEntry {
-    uint32_t offset;     // offset in the message object
-    int32_t has_idx;     // has-bit index, relative to the message object
-    uint16_t aux_idx;    // index for `field_aux`.
-    uint16_t type_card;  // `FieldType` and `Cardinality` (see _impl.h)
+    constexpr FieldEntry() : type_card(0), has_idx(0), aux_idx(0), offset(0) {}
+    constexpr FieldEntry(uint32_t o, int32_t h, uint16_t a, uint16_t t) : type_card(t), has_idx(h), aux_idx(a), offset(o) {
+      // Prevent silent overflows
+      if (o >= (1 << 18)) NotConstexpr();
+      if (h >= (1 << 16)) NotConstexpr();
+      if (h < -1) NotConstexpr();
+      if (a >= (1 << 14)) NotConstexpr();
+    }
+    uint64_t type_card : 16;  // `FieldType` and `Cardinality` (see _impl.h)
+    uint64_t has_idx : 16;     // has-bit index, relative to the message object
+    uint64_t aux_idx : 14;    // index for `field_aux`.
+    uint64_t offset : 18;     // offset in the message object
 
     static constexpr uint16_t kNoAuxIdx = 0xFFFF;
```

Comment:
If aux_idx becomes 14 bits instead of 16bits, kNoAuxIdx needs to be changed as well, otherwise, internal check against kNoAuxIdx fails. A temp fix is to make it 0x3FFFF.

Reply:
That makes a lot of sense